### PR TITLE
feat: collectible, show get more token button when balance insufficient 

### PR DESCRIPTION
**Pull request by BillySh1**
## Description

Closes # (MASK-522 MASK-537)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Previews

image
when clicked get more eth， call the swap dialog
image

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [x] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [x] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

**Comment by BillySh1** [comment]
@DimensionDev/qa  cc

File: packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx [hunk 1]
```diff
@@ -71,7 +72,11 @@ export function CheckoutDialog(props: CheckoutDialogProps) {
 
     const [unreviewedChecked, setUnreviewedChecked] = useState(false)
     const [ToS_Checked, setToS_Checked] = useState(false)
-
+    const [is_insufficient_balance, setIs_insufficient_balance] = useState(false)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx:75]: We prefer camelCased variable names.

File: packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx [hunk 2]
```diff
@@ -84,13 +89,35 @@ export function CheckoutDialog(props: CheckoutDialogProps) {
...
+                    decimals: token.value.decimals,
+                } as Coin,
+            },
+        })
+    }, [token.value, openSwapDialog])
 
     const validationMessage = useMemo(() => {
+        const basePrice = new BigNumber(order?.value?.basePrice ?? '0')
+        const balance_ = new BigNumber(balance.value ?? '0')
+        if (basePrice.isGreaterThan(balance_)) {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx:113]: Use `isGreaterThan` to replace it.

File: packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutOrder.tsx [hunk 3]
```diff
@@ -24,14 +27,24 @@ export interface CheckoutOrderProps {
...
     if (!order) return null
 
+    const price = (order as Order).currentPrice ?? asset.value.current_price
+    const getPrice = () => {
+        if (!price) return 'error'
+        const decimal = asset.value?.response_.collection.payment_tokens.find((item: ERC20TokenDetailed) => {
+            return item.symbol === asset.value?.current_symbol
+        })?.decimals
+        if (!decimal) return 'error'
+        return leftShift(new BigNumber(price), decimal).toFixed() ?? 'error'
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutOrder.tsx:45]: Why not use `formatBalance` here?

File: packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx [hunk 4]
```diff
@@ -91,6 +91,7 @@ export function MakeOfferDialog(props: MakeOfferDialogProps) {
     const [expirationDateTime, setExpirationDateTime] = useState(new Date())
     const [unreviewedChecked, setUnreviewedChecked] = useState(false)
     const [ToS_Checked, setToS_Checked] = useState(false)
+    const [is_insufficient_balance, setIs_insufficient_balance] = useState(false)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx:94]: Also here.

File: packages/mask/src/plugins/Collectible/hooks/useAssetOrder.ts [hunk 5]
```diff
@@ -16,13 +16,12 @@ export function useAssetOrder(provider: NonFungibleAssetProvider, token?: Collec
                 const openSeaResponse = await PluginCollectibleRPC.getAssetFromSDK(token.contractAddress, token.tokenId)
                 const getPrice = (order: Order) =>
                     getOrderUnitPrice(
-                        order.currentPrice?.toFixed(),
+                        order.currentPrice as unknown as string,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Collectible/hooks/useAssetOrder.ts:19]: As I can see from the code, the original type is BigNumber. Does the latest version of opensae-js fix this issue?

image

File: packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx (hunk 1 above)
Comment (BillySh1) [reply packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx:75]: sure

File: packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutOrder.tsx (hunk 3 above)
Comment (BillySh1) [reply packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutOrder.tsx:45]: repalced with formatBalance()

File: packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx (hunk 2 above)
Comment (septs) [reply packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx:113]:
```suggestion
        if (isGreaterThan(basePrice, balance.value ?? 0)) {
```

File: packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx (hunk 4 above)
Comment (septs) [reply packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx:94]:
```suggestion
    const [isInsufficientBalance, setInsufficientBalance] = useState(false)
```

File: packages/mask/src/plugins/Collectible/hooks/useAssetOrder.ts (hunk 5 above)
Comment (BillySh1) [reply packages/mask/src/plugins/Collectible/hooks/useAssetOrder.ts]: Up to opensea-js latest version 1.2.7, type still BigNumber when the response type is string, assert type as AssetOrder and its works fine

File: packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx [hunk 6]
```diff
@@ -67,12 +67,11 @@ export function CheckoutDialog(props: CheckoutDialogProps) {
     const isVerified = asset?.value?.is_verified ?? false
     const { t } = useI18N()
     const { classes } = useStyles()
-
     const account = useAccount()
 
     const [unreviewedChecked, setUnreviewedChecked] = useState(false)
     const [ToS_Checked, setToS_Checked] = useState(false)
-    const [is_insufficient_balance, setIs_insufficient_balance] = useState(false)
+    const [isInsufficientBalance, setIsInsufficientBalance] = useState(false)
```
Comment (septs) [inline packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutDialog.tsx]:
```suggestion
    const [isInsufficientBalance, setInsufficientBalance] = useState(false)
```

File: packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutOrder.tsx [hunk 7]
```diff
@@ -24,14 +26,23 @@ export interface CheckoutOrderProps {
...
     if (!order) return null
 
+    const price = (order as Order).currentPrice ?? asset.value.current_price
+    const getPrice = () => {
+        if (!price) return 'error'
+        const decimal = asset.value?.response_.collection.payment_tokens.find((item: ERC20TokenDetailed) => {
+            return item.symbol === asset.value?.current_symbol
+        })?.decimals
+        if (!decimal) return 'error'
+        return formatBalance(new BigNumber(price), decimal) ?? 'error'
```
Comment (septs) [inline packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutOrder.tsx]:
```suggestion
        return formatBalance(price, decimal) ?? 'error'
```

File: packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx (hunk 4 above)
Comment (septs) [reply packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx]: rename variable name.

**Comment by orange187** [comment]
comment

**Comment by orange187** [comment]
Checked

File: packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutOrder.tsx (hunk 7 above)
Comment (BillySh1) [reply packages/mask/src/plugins/Collectible/SNSAdaptor/CheckoutOrder.tsx]: modified

File: packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx (hunk 4 above)
Comment (BillySh1) [reply packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx]: renamed

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**